### PR TITLE
Add weaver-graph and weaver-e2e crates; LSP call hierarchy support

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Introduce a new weaver-graph crate that models a bidirectional call graph with a provider-based architecture, starting with an LSP-based provider. Also add an end-to-end test harness (weaver-e2e) and extend the LSP host scaffolding to support call hierarchy related capabilities. This lays groundwork for rich call graph construction from multiple data sources (LSP, static, dynamic) and includes unit tests for graph components and e2e tests for integration with a Pyrefly-based Python language server.
- Extend workspace to include weaver-e2e and weaver-graph crates and update workspace wiring accordingly.

## Changes
- Add new crate: weaver-graph
  - Core data structures for call graphs:
    - node.rs: CallNode, NodeId, SymbolKind, and node helpers
    - edge.rs: CallEdge, EdgeSource, and call-site metadata
    - graph.rs: CallGraph with bidirectional indexing (incoming/outgoing) and traversal helpers
  - Error handling: error.rs with GraphError and helper constructors
  - Provider framework: provider.rs
    - SourcePosition orients queries from a specific source location
    - CallGraphProvider trait for generic graph construction
    - CallHierarchyClient trait to abstract LSP call hierarchy interactions
    - LspCallGraphProvider: LSP-backed provider wrapper
  - lib.rs: Public API exports and documentation
  - tests.rs: Unit tests covering graph, nodes, and edges
  - uri.rs: URI/path conversion utilities
- Add end-to-end testing crate: weaver-e2e
  - Cargo manifest and test dependencies wired to weaver-graph and weaver-lsp-host
  - lsp_client.rs: Generic LSP client for E2E scenarios
  - pyrefly.rs: Pyrefly-specific LSP client wrapper
  - fixtures.rs: Python code samples for call graph scenarios (linear chain, diamond, class methods, no calls)
  - lib.rs: E2E test harness overview and utilities
  - tests/call_hierarchy.rs: E2E tests exercising prepareCallHierarchy, incomingCalls, and outgoingCalls with Pyrefly
- Extend weaver-lsp-host to support call hierarchy capabilities
  - capability.rs: Introduce CallHierarchy capability kind and propagate capability state
  - doc_support.rs: Stubbed/extended support to document prepareCallHierarchy/incomingCalls/outgoingCalls usage
  - host.rs: Extend LSP host with methods for prepare_call_hierarchy, incoming_calls, outgoing_calls
  - server.rs: Extend server capability/trait definitions to include call hierarchy methods
  - Tests and support scaffolding updated to exercise call hierarchy paths
- Workspace integration
  - Cargo.toml: add weaver-e2e and weaver-graph to workspace members
  - Cargo.lock updated accordingly

## Core Components
- weaver_graph crate
  - Node model: unique NodeId, name, kind, path, position, optional container, qualified name
  - Edge model: CallEdge with caller, callee, provenance (EdgeSource), optional call-site location
  - Graph: CallGraph maintaining nodes, edges, and bidirectional indices for fast queries (incoming/outgoing edges, callers/callees)
  - Errors: GraphError with common failure cases and helpers
  - Providers: trait-based provider system to fuse data sources; LSP provider implemented and ready for integration
- weaver-e2e
  - End-to-end infrastructure to validate Weaver components against a real language server (Pyrefly) including LSP operations around call hierarchy
  - Fixtures for Python code demonstrating various call graph shapes
  - Pyrefly LSP client wrapper and a simple LSP client abstraction
  - Tests covering prepare_call_hierarchy, outgoing/incoming calls, and edge cases (e.g., no calls)
- weaver-lsp-host
  - Expanded capability model to track CallHierarchy support and surface prepareCallHierarchy/incomingCalls/outgoingCalls through the language server protocol
  - Tests/utilities updated to accommodate call hierarchy behavior in multiple languages

## API Overview
- weaver_graph::CallGraph provides core graph semantics with:
  - add_node, add_edge, node, nodes, edges, incoming_edges, outgoing_edges
  - callers_of, callees_of, find_by_name, get_node, merge, is_empty
- Provider abstractions offer:
  - CallGraphProvider: build_graph, callers_graph, callees_graph
  - CallHierarchyClient: prepare_call_hierarchy, incoming_calls, outgoing_calls
  - LspCallGraphProvider: ready to be constructed with any CallHierarchyClient implementation

## Testing Plan
- Unit tests for weaver-graph (nodes, edges, graph traversal, find_by_name, qualified_name handling)
- E2E tests in weaver-e2e for call hierarchy scenarios against Pyrefly LSP server
  - Tests guarded to gracefully skip if Pyrefly is not available (uvx-based installation)
- LSP host tests extended to validate the call hierarchy capability surface

## How to Run
- Unit tests for graph crate only:
  - cargo test -p weaver-graph
- Run end-to-end tests (requires uvx and Pyrefly tooling):
  - cargo test -p weaver-e2e
- Run the full workspace tests:
  - cargo test --workspace

## Notes & Considerations
- This patch introduces scaffolding and interfaces for call hierarchy data sources. Real data population from a concrete language server may require enabling additional features or providing concrete implementations of CallHierarchyClient.
- E2E tests depend on external tooling (e.g., uvx, Pyrefly). Tests are designed to skip gracefully when these tools are not present to keep CI robust.
- The LSP host capability surface is extended to advertise and respond to call hierarchy capabilities, aligning with forthcoming features in language servers.

If you want me to tailor the PR body toward a specific reviewer or team (e.g., “graph team” vs “lsp integration team”), I can adjust the sections accordingly.

🌿 Generated by [Terry](https://www.terragonlabs.com)

---

ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/633f4c30-ae0f-41d7-bbed-83be650e8e7f

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-graph/src/node.rs](https://github.com/leynos/weaver/pull/41/files/de02eec13e9383e38b0fee9a5ac04608fae14f5e#diff-07f8074855f5bc308aee77fe11b491b7c31c60a7e4af94a6428c4a0192fdb1e4)

Comment on lines +89 to +108

~~~rust
    pub fn new(
        name: impl Into<String>,
        kind: SymbolKind,
        path: impl Into<Utf8PathBuf>,
        line: u32,
        column: u32,
    ) -> Self {
        let name_str = name.into();
        let path_buf = path.into();
        let id = NodeId::new(&path_buf, line, column, &name_str);
        Self {
            id,
            name: name_str,
            kind,
            path: path_buf,
            line,
            column,
            container: None,
        }
    }
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=41&biomarker=Excess+Number+of+Function+Arguments&filename=crates%2Fweaver-graph%2Fsrc%2Fnode.rs&method=CallNode.new)
CallNode.new has 5 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-graph/src/provider.rs](https://github.com/leynos/weaver/pull/41/files/de02eec13e9383e38b0fee9a5ac04608fae14f5e#diff-9f33f92cb3dd4532cb8cf8b97ed30af762651cef36e749f5706f687f8dbf2aea)

Comment on lines +177 to +201

~~~rust
    fn callers_graph(
        &mut self,
        position: &SourcePosition,
        depth: u32,
    ) -> Result<CallGraph, GraphError> {
        let mut graph = CallGraph::new();

        let items = self.prepare_at_position(position)?;

        if items.is_empty() {
            return Err(GraphError::symbol_not_found(
                &position.path,
                position.line,
                position.column,
            ));
        }

        for item in &items {
            let node = call_hierarchy_item_to_node(item);
            graph.add_node(node);
            self.explore_callers(&mut graph, item, depth)?;
        }

        Ok(graph)
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=41&biomarker=Code+Duplication&filename=crates%2Fweaver-graph%2Fsrc%2Fprovider.rs&method=)
The module contains 4 functions with similar structure: CallGraphProvider.callees_graph,CallGraphProvider.callers_graph,LspCallGraphProvider.explore_callees,LspCallGraphProvider.explore_callers

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
crates/weaver-graph/src/provider.rs: 

### What lead to degradation?
The module contains 4 functions with similar structure: CallGraphProvider.callees_graph,CallGraphProvider.callers_graph,LspCallGraphProvider.explore_callees,LspCallGraphProvider.explore_callers

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-graph/src/uri.rs](https://github.com/leynos/weaver/pull/41/files/cb76bd04baf63310cb02065c1136831eb39a6445#diff-4e9ba78223a4422ade79bd305918cb0cd15af1c367431e439ed80c623df0049e)

Comment on lines +20 to +22

~~~rust
    if let Ok(url) = Url::parse(uri_str)
        && let Ok(path) = url.to_file_path()
        && let Ok(utf8_path) = Utf8PathBuf::try_from(path)
~~~

❌ New issue: [Complex Conditional](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=41&biomarker=Complex+Conditional&filename=crates%2Fweaver-graph%2Fsrc%2Furi.rs&method=uri_to_path)
uri_to_path has 1 complex conditionals with 2 branches, threshold = 2

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-e2e/src/lsp_client.rs](https://github.com/leynos/weaver/pull/41/files/cb76bd04baf63310cb02065c1136831eb39a6445#diff-7e1b43e0ff0fea9e87aab8438e5d4bb5568089c868fa026876f04d8e425fbfe6)

Comment on lines +199 to +208

~~~rust
    pub fn prepare_call_hierarchy(
        &mut self,
        params: CallHierarchyPrepareParams,
    ) -> Result<Option<Vec<CallHierarchyItem>>, LspClientError> {
        self.require_initialized()?;
        self.request(
            "textDocument/prepareCallHierarchy",
            Some(serde_json::to_value(params).map_err(LspClientError::Json)?),
        )
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=41&biomarker=Code+Duplication&filename=crates%2Fweaver-e2e%2Fsrc%2Flsp_client.rs&method=)
The module contains 3 functions with similar structure: LspClient.incoming_calls,LspClient.outgoing_calls,LspClient.prepare_call_hierarchy

File: crates/weaver-graph/src/provider.rs [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+            // If decoding failed, keep the original
+            result.push('%');
+            result.push_str(&hex);
+        } else {
+            result.push(c);
+        }
+    }
+
+    result
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-graph/src/provider.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: crates/weaver-graph/src/provider.rs [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+/// Converts a path to a file:// URI.
+fn path_to_uri(path: &Utf8PathBuf) -> Result<Uri, GraphError> {
+    let uri_string = format!("file://{path}");
+    uri_string.parse().map_err(|_| {
+        GraphError::io(
+            format!("failed to convert path to URI: {path}"),
+            std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid path"),
+        )
+    })
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-graph/src/provider.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: crates/weaver-e2e/tests/call_hierarchy.rs [hunk 3]
```diff
@@ -0,0 +1,256 @@
+//! End-to-end tests for call hierarchy functionality.
+//!
+//! These tests exercise the call hierarchy features against a real Pyrefly
+//! language server. Tests are skipped gracefully if Pyrefly is not available.
+
+#![expect(clippy::expect_used, reason = "test code uses expect for assertions")]
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-e2e/tests/call_hierarchy.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: crates/weaver-graph/src/edge.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            call_site: None,
+        }
+    }
+
+    /// Sets the call site location.
+    #[must_use]
+    pub const fn with_call_site(mut self, line: u32, column: u32) -> Self {
+        self.call_site = Some(Position::new(line, column));
+        self
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-graph/src/edge.rs:60]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: crates/weaver-graph/src/uri.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    // Try to parse as a URL and extract the file path
+    if let Some(path) = try_parse_uri_to_path(uri_str) {
+        return path;
+    }
+
+    // Fallback: strip file:// prefix manually
+    uri_str
+        .strip_prefix("file://")
+        .map_or_else(|| Utf8PathBuf::from(uri_str), Utf8PathBuf::from)
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-graph/src/uri.rs:35]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: crates/weaver-graph/src/uri.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        )
+    })?;
+
+    url.as_str().parse().map_err(|_| {
+        GraphError::io(
+            format!("failed to parse URI: {}", url.as_str()),
+            std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid URI"),
+        )
+    })
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-graph/src/uri.rs:57]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-e2e/tests/call_hierarchy.rs](https://github.com/leynos/weaver/pull/41/files/fd4ebab5b87626cb6ee3287e60f7ec849e5f0474#diff-5fb1772f86b892dfa0b892bbb692d2f2ba4d68661ec5792fdf988a616fb3068d)

Comment on lines +56 to +79

~~~rust
    pub fn linear_chain_context() -> Option<TestContext> {
        if !pyrefly_available() {
            return None;
        }

        let temp_dir = TempDir::new().expect("create temp dir");
        let file_path = temp_dir.path().join("test.py");
        std::fs::write(&file_path, fixtures::LINEAR_CHAIN).expect("write test file");

        let root_uri = file_uri(temp_dir.path());
        let file_uri_val = file_uri(&file_path);

        let mut client = LspClient::spawn("uvx", &["pyrefly", "lsp"]).expect("spawn pyrefly");
        client.initialize(root_uri).expect("initialize");
        client
            .did_open(file_uri_val.clone(), "python", fixtures::LINEAR_CHAIN)
            .expect("open file");

        Some(TestContext {
            client,
            file_uri: file_uri_val,
            _temp_dir: temp_dir,
        })
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=41&biomarker=Code+Duplication&filename=crates%2Fweaver-e2e%2Ftests%2Fcall_hierarchy.rs&method=)
The module contains 4 functions with similar structure: fixtures_impl.linear_chain_context,fixtures_impl.no_calls_context,test_impl.incoming_calls_returns_callers_impl,test_impl.outgoing_calls_returns_callees_impl

File: crates/weaver-e2e/src/pyrefly.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    /// # Errors
+    /// Returns an error if the notification fails.
+    pub fn open_python_file(&mut self, uri: Uri, text: &str) -> Result<(), LspClientError> {
+        self.client.did_open(uri, "python", text)
+    }
+
+    /// Returns a mutable reference to the underlying LSP client.
+    pub const fn client_mut(&mut self) -> &mut LspClient {
+        &mut self.client
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-e2e/src/pyrefly.rs:45]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: crates/weaver-e2e/tests/call_hierarchy.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-e2e/tests/call_hierarchy.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-e2e/tests/call_hierarchy.rs](https://github.com/leynos/weaver/pull/41/files/d00c61e6a3ccc9980aaeb81425041bcd98afd07d#diff-5fb1772f86b892dfa0b892bbb692d2f2ba4d68661ec5792fdf988a616fb3068d)

Comment on lines +129 to +172

~~~rust
    pub fn outgoing_calls_returns_callees_impl(ctx: &mut TestContext) {
        // First prepare call hierarchy for function `a`
        let prepare_params = CallHierarchyPrepareParams {
            text_document_position_params: TextDocumentPositionParams {
                text_document: TextDocumentIdentifier {
                    uri: ctx.file_uri.clone(),
                },
                position: Position::new(0, 4),
            },
            work_done_progress_params: WorkDoneProgressParams::default(),
        };

        let items = ctx
            .client
            .prepare_call_hierarchy(prepare_params)
            .expect("prepare")
            .expect("items");

        let item = items.into_iter().next().expect("first item");

        // Now get outgoing calls
        let outgoing_params = CallHierarchyOutgoingCallsParams {
            item,
            work_done_progress_params: WorkDoneProgressParams::default(),
            partial_result_params: lsp_types::PartialResultParams::default(),
        };

        let outgoing = ctx
            .client
            .outgoing_calls(outgoing_params)
            .expect("outgoing calls");

        let calls = outgoing.expect("should have outgoing calls");
        assert!(!calls.is_empty(), "should have at least one call");

        // `a` calls `b`
        let callee_names: Vec<_> = calls.iter().map(|c| c.to.name.as_str()).collect();
        assert!(
            callee_names.contains(&"b"),
            "should include call to `b`, got: {callee_names:?}"
        );

        ctx.client.shutdown().expect("shutdown");
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=41&biomarker=Code+Duplication&filename=crates%2Fweaver-e2e%2Ftests%2Fcall_hierarchy.rs&method=)
The module contains 2 functions with similar structure: test_impl.incoming_calls_returns_callers_impl,test_impl.outgoing_calls_returns_callees_impl

File: crates/weaver-e2e/src/pyrefly.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-e2e/src/pyrefly.rs:45]: @coderabbitai The coderabbitai analysis is incorrect for nightly Rust. The code compiles successfully with const fn and mutable references, and removing const would fail the clippy::missing_const_for_fn lint. This is a false positive from the reviewer.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-e2e/tests/call_hierarchy.rs](https://github.com/leynos/weaver/pull/41/files/a40380418b74eb04b8b24cc8f4a78faddb2acc03#diff-5fb1772f86b892dfa0b892bbb692d2f2ba4d68661ec5792fdf988a616fb3068d)

Comment on lines +165 to +189

~~~rust
    fn assert_outgoing_calls_contain(
        ctx: &mut TestContext,
        item: CallHierarchyItem,
        expected_callee: &str,
    ) {
        let outgoing_params = CallHierarchyOutgoingCallsParams {
            item,
            work_done_progress_params: WorkDoneProgressParams::default(),
            partial_result_params: lsp_types::PartialResultParams::default(),
        };

        let calls = ctx
            .client
            .outgoing_calls(outgoing_params)
            .expect("outgoing calls")
            .expect("should have outgoing calls");

        assert!(!calls.is_empty(), "should have at least one call");

        let callee_names: Vec<_> = calls.iter().map(|c| c.to.name.as_str()).collect();
        assert!(
            callee_names.contains(&expected_callee),
            "should include call to `{expected_callee}`, got: {callee_names:?}"
        );
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=41&biomarker=Code+Duplication&filename=crates%2Fweaver-e2e%2Ftests%2Fcall_hierarchy.rs&method=)
The module contains 2 functions with similar structure: test_impl.assert_incoming_calls_contain,test_impl.assert_outgoing_calls_contain

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review
